MueLu: warnings from catching polymorphic types by value

## Bug Report
@trilinos/muelu

### Description

Many ``-Wcatch-value`` warnings caused by catching polymorphic types (subclasses of std::exception) by value, instead of reference. Example:

```
/ascldap/users/bmkelle/MiscWork/Trilinos/packages/muelu/src/Interface/../Utils/MueLu_Utilities_decl.hpp:363:23: warning: catching polymorphic type ‘class std::bad_cast’ by value [-Wcatch-value=]
  363 |         } catch (std::bad_cast) {
```

### Steps to Reproduce
Built from current develop branch with GCC 9.1.

--- a/packages/muelu/adapters/epetra/MueLu_CreateEpetraPreconditioner.cpp
+++ b/packages/muelu/adapters/epetra/MueLu_CreateEpetraPreconditioner.cpp
@@ -52,7 +52,7 @@ namespace MueLu {
       RCP<Xpetra::MultiVector<typename Teuchos::ScalarTraits<SC>::coordinateType,LO,GO,NO> > coordinates = Teuchos::null;
       try {
         coordinates = EpetraMultiVector_To_XpetraMultiVector<typename Teuchos::ScalarTraits<SC>::coordinateType,LO,GO,NO>(userList.get<RCP<Epetra_MultiVector> >("Coordinates"));
-      } catch(Teuchos::Exceptions::InvalidParameterType) {
+      } catch(Teuchos::Exceptions::InvalidParameterType&) {
         coordinates = userList.get<RCP<Xpetra::MultiVector<typename Teuchos::ScalarTraits<SC>::coordinateType, LO, GO, NO> > >("Coordinates");
       }
       if(Teuchos::nonnull(coordinates)){
@@ -63,7 +63,7 @@ namespace MueLu {
       RCP<Xpetra::MultiVector<typename Teuchos::ScalarTraits<SC>::coordinateType,LO,GO,NO> > nullspace = Teuchos::null;
       try {
         nullspace = EpetraMultiVector_To_XpetraMultiVector<SC,LO,GO,NO>(userList.get<RCP<Epetra_MultiVector> >("Nullspace"));
-      } catch(Teuchos::Exceptions::InvalidParameterType) {
+      } catch(Teuchos::Exceptions::InvalidParameterType&) {
         nullspace = userList.get<RCP<Xpetra::MultiVector<SC, LO, GO, NO> > >("Nullspace");
       }
       if(Teuchos::nonnull(nullspace)){

--- a/packages/muelu/adapters/tpetra/MueLu_CreateTpetraPreconditioner.hpp
+++ b/packages/muelu/adapters/tpetra/MueLu_CreateTpetraPreconditioner.hpp
@@ -89,7 +89,7 @@ namespace MueLu {
       RCP<Xpetra::MultiVector<typename Teuchos::ScalarTraits<Scalar>::coordinateType,LO,GO,NO> > coordinates = Teuchos::null;
       try {
         coordinates = TpetraMultiVector_To_XpetraMultiVector<typename Teuchos::ScalarTraits<Scalar>::coordinateType,LO,GO,NO>(userList.get<RCP<Tpetra::MultiVector<typename Teuchos::ScalarTraits<Scalar>::coordinateType, LocalOrdinal, GlobalOrdinal, Node> > >("Coordinates"));
-      } catch(Teuchos::Exceptions::InvalidParameterType) {
+      } catch(Teuchos::Exceptions::InvalidParameterType&) {
         coordinates = userList.get<RCP<Xpetra::MultiVector<typename Teuchos::ScalarTraits<Scalar>::coordinateType, LocalOrdinal, GlobalOrdinal, Node> > >("Coordinates");
       }
       userList.set<RCP<Xpetra::MultiVector<typename Teuchos::ScalarTraits<Scalar>::coordinateType,LO,GO,NO> > >("Coordinates", coordinates);
@@ -99,7 +99,7 @@ namespace MueLu {
       RCP<MultiVector> nullspace = Teuchos::null;
       try {
         nullspace = TpetraMultiVector_To_XpetraMultiVector<SC,LO,GO,NO>(userList.get<RCP<Tpetra::MultiVector<Scalar, LocalOrdinal, GlobalOrdinal, Node> > >("Nullspace"));
-      } catch(Teuchos::Exceptions::InvalidParameterType) {
+      } catch(Teuchos::Exceptions::InvalidParameterType&) {
         nullspace = userList.get<RCP<Xpetra::MultiVector<Scalar, LocalOrdinal, GlobalOrdinal, Node> > >("Nullspace");
       }
       userList.set<RCP<MultiVector> >("Nullspace", nullspace);

--- a/packages/muelu/src/Graph/HybridAggregation/MueLu_HybridAggregationFactory_def.hpp
+++ b/packages/muelu/src/Graph/HybridAggregation/MueLu_HybridAggregationFactory_def.hpp
@@ -301,7 +301,7 @@ namespace MueLu {
     Teuchos::Array<LO> coarseRate;
     try {
       coarseRate = Teuchos::fromStringToArray<LO>(coarseningRate);
-    } catch(const Teuchos::InvalidArrayStringRepresentation e) {
+    } catch(const Teuchos::InvalidArrayStringRepresentation& e) {
       GetOStream(Errors,-1) << " *** \"aggregation: coarsening rate\" must be a string convertible into an array! *** "
                             << std::endl;
       throw e;

--- a/packages/muelu/src/Graph/StructuredAggregation/MueLu_StructuredAggregationFactory_def.hpp
+++ b/packages/muelu/src/Graph/StructuredAggregation/MueLu_StructuredAggregationFactory_def.hpp
@@ -214,7 +214,7 @@ namespace MueLu {
     Teuchos::Array<LO> coarseRate;
     try {
       coarseRate = Teuchos::fromStringToArray<LO>(coarseningRate);
-    } catch(const Teuchos::InvalidArrayStringRepresentation e) {
+    } catch(const Teuchos::InvalidArrayStringRepresentation& e) {
       GetOStream(Errors,-1) << " *** \"aggregation: coarsening rate\" must be a string convertible into an array! *** "
                             << std::endl;
       throw e;

--- a/packages/muelu/src/Graph/StructuredAggregation/MueLu_StructuredAggregationFactory_kokkos_def.hpp
+++ b/packages/muelu/src/Graph/StructuredAggregation/MueLu_StructuredAggregationFactory_kokkos_def.hpp
@@ -189,7 +189,7 @@ namespace MueLu {
     Teuchos::Array<LO> coarseRate;
     try {
       coarseRate = Teuchos::fromStringToArray<LO>(coarseningRate);
-    } catch(const Teuchos::InvalidArrayStringRepresentation e) {
+    } catch(const Teuchos::InvalidArrayStringRepresentation& e) {
       GetOStream(Errors,-1) << " *** \"aggregation: coarsening rate\" must be a string convertible into an array! *** "
                             << std::endl;
       throw e;

--- a/packages/muelu/src/Interface/MueLu_ParameterListInterpreter_def.hpp
+++ b/packages/muelu/src/Interface/MueLu_ParameterListInterpreter_def.hpp
@@ -211,7 +211,7 @@ namespace MueLu {
     if      (paramList  .isParameter(paramName)) listWrite.set(paramName, paramList  .get<paramType>(paramName)); \
     else if (defaultList.isParameter(paramName)) listWrite.set(paramName, defaultList.get<paramType>(paramName)); \
   } \
-  catch(Teuchos::Exceptions::InvalidParameterType) { \
+  catch(Teuchos::Exceptions::InvalidParameterType&) { \
     TEUCHOS_TEST_FOR_EXCEPTION_PURE_MSG(true, Teuchos::Exceptions::InvalidParameterType, \
                                         "Error: parameter \"" << paramName << "\" must be of type " << Teuchos::TypeNameTraits<paramType>::name()); \
   } \
@@ -1089,7 +1089,7 @@ namespace MueLu {
         RAPparams.set("RepairMainDiagonal", defaultList.get<bool>("aggregation: allow empty prolongator columns"));
       }
 
-    } catch (Teuchos::Exceptions::InvalidParameterType) {
+    } catch (Teuchos::Exceptions::InvalidParameterType&) {
       TEUCHOS_TEST_FOR_EXCEPTION_PURE_MSG(true, Teuchos::Exceptions::InvalidParameterType,
           "Error: parameter \"aggregation: allow empty prolongator columns\" must be of type " << Teuchos::TypeNameTraits<bool>::name());
     }

--- a/packages/muelu/src/Smoothers/MueLu_Ifpack2Smoother_def.hpp
+++ b/packages/muelu/src/Smoothers/MueLu_Ifpack2Smoother_def.hpp
@@ -236,7 +236,7 @@ namespace MueLu {
       RCP<const tRowMatrix> tA;
       try {
         tA = Utilities::Op2NonConstTpetraRow(A_);
-      } catch (Exceptions::BadCast) {
+      } catch (Exceptions::BadCast&) {
         isTRowMatrix = false;
       }
 

--- a/packages/muelu/src/Smoothers/MueLu_IfpackSmoother.cpp
+++ b/packages/muelu/src/Smoothers/MueLu_IfpackSmoother.cpp
@@ -141,7 +141,7 @@ namespace MueLu {
         lambdaMax = Teuchos::getValue<Scalar>(this->GetParameter(maxEigString));
         this->GetOStream(Statistics1) << maxEigString << " (cached with smoother parameter list) = " << lambdaMax << std::endl;
 
-      } catch (Teuchos::Exceptions::InvalidParameterName) {
+      } catch (Teuchos::Exceptions::InvalidParameterName&) {
         lambdaMax = A_->GetMaxEigenvalueEstimate();
 
         if (lambdaMax != -1.0) {
@@ -157,7 +157,7 @@ namespace MueLu {
       try {
         ratio = Teuchos::getValue<Scalar>(this->GetParameter(eigRatioString));
 
-      } catch (Teuchos::Exceptions::InvalidParameterName) {
+      } catch (Teuchos::Exceptions::InvalidParameterName&) {
         this->SetParameter(eigRatioString, ParameterEntry(ratio));
       }
 

--- a/packages/muelu/src/Transfers/BlackBox/MueLu_BlackBoxPFactory_def.hpp
+++ b/packages/muelu/src/Transfers/BlackBox/MueLu_BlackBoxPFactory_def.hpp
@@ -187,7 +187,7 @@ namespace MueLu {
       Teuchos::Array<LO> crates;
       try {
         crates = Teuchos::fromStringToArray<LO>(coarsenRate);
-      } catch(const Teuchos::InvalidArrayStringRepresentation e) {
+      } catch(const Teuchos::InvalidArrayStringRepresentation& e) {
         GetOStream(Errors,-1) << " *** Coarsen must be a string convertible into an array! *** "
                               << std::endl;
         throw e;

--- a/packages/muelu/src/Transfers/GeneralGeometric/MueLu_GeneralGeometricPFactory_def.hpp
+++ b/packages/muelu/src/Transfers/GeneralGeometric/MueLu_GeneralGeometricPFactory_def.hpp
@@ -217,7 +217,7 @@ namespace MueLu {
     Teuchos::Array<LO> crates;
     try {
       crates = Teuchos::fromStringToArray<LO>(coarsenRate);
-    } catch(const Teuchos::InvalidArrayStringRepresentation e) {
+    } catch(const Teuchos::InvalidArrayStringRepresentation& e) {
       GetOStream(Errors,-1) << " *** Coarsen must be a string convertible into an array! *** "
                             << std::endl;
       throw e;

--- a/packages/muelu/src/Utils/MueLu_TimeMonitor.cpp
+++ b/packages/muelu/src/Utils/MueLu_TimeMonitor.cpp
@@ -89,7 +89,7 @@ namespace MueLu {
         if (nonnull(stackedTimer))
           stackedTimer->stop(timer_->name());
       }
-      catch (std::runtime_error) {
+      catch (std::runtime_error&) {
         std::ostringstream warning;
         warning <<
           "\n*********************************************************************\n"

--- a/packages/muelu/src/Utils/MueLu_Utilities_decl.hpp
+++ b/packages/muelu/src/Utils/MueLu_Utilities_decl.hpp
@@ -360,10 +360,10 @@ namespace MueLu {
         try {
           const EpetraCrsMatrix& tmp_ECrsMtx = dynamic_cast<const EpetraCrsMatrix&>(*crsOp.getCrsMatrix());
           return *tmp_ECrsMtx.getEpetra_CrsMatrix();
-        } catch (std::bad_cast) {
+        } catch (std::bad_cast&) {
           throw Exceptions::BadCast("Cast from Xpetra::CrsMatrix to Xpetra::EpetraCrsMatrix failed");
         }
-      } catch (std::bad_cast) {
+      } catch (std::bad_cast&) {
         throw Exceptions::BadCast("Cast from Xpetra::Matrix to Xpetra::CrsMatrixWrap failed");
       }
     }
@@ -373,10 +373,10 @@ namespace MueLu {
         try {
           EpetraCrsMatrix& tmp_ECrsMtx = dynamic_cast<EpetraCrsMatrix&>(*crsOp.getCrsMatrix());
           return *tmp_ECrsMtx.getEpetra_CrsMatrixNonConst();
-        } catch (std::bad_cast) {
+        } catch (std::bad_cast&) {
           throw Exceptions::BadCast("Cast from Xpetra::CrsMatrix to Xpetra::EpetraCrsMatrix failed");
         }
-      } catch (std::bad_cast) {
+      } catch (std::bad_cast&) {
         throw Exceptions::BadCast("Cast from Xpetra::Matrix to Xpetra::CrsMatrixWrap failed");
       }
     }
@@ -484,10 +484,10 @@ namespace MueLu {
         try {
           const Xpetra::TpetraCrsMatrix<Scalar,LocalOrdinal,GlobalOrdinal,Node>& tmp_ECrsMtx = dynamic_cast<const Xpetra::TpetraCrsMatrix<Scalar,LocalOrdinal,GlobalOrdinal,Node>&>(*crsOp.getCrsMatrix());
           return *tmp_ECrsMtx.getTpetra_CrsMatrix();
-        } catch (std::bad_cast) {
+        } catch (std::bad_cast&) {
           throw Exceptions::BadCast("Cast from Xpetra::CrsMatrix to Xpetra::TpetraCrsMatrix failed");
         }
-      } catch (std::bad_cast) {
+      } catch (std::bad_cast&) {
         throw Exceptions::BadCast("Cast from Xpetra::Matrix to Xpetra::CrsMatrixWrap failed");
       }
 #endif
@@ -502,10 +502,10 @@ namespace MueLu {
         try {
           Xpetra::TpetraCrsMatrix<Scalar,LocalOrdinal,GlobalOrdinal,Node>& tmp_ECrsMtx = dynamic_cast<Xpetra::TpetraCrsMatrix<Scalar,LocalOrdinal,GlobalOrdinal,Node>&>(*crsOp.getCrsMatrix());
           return *tmp_ECrsMtx.getTpetra_CrsMatrixNonConst();
-        } catch (std::bad_cast) {
+        } catch (std::bad_cast&) {
           throw Exceptions::BadCast("Cast from Xpetra::CrsMatrix to Xpetra::TpetraCrsMatrix failed");
         }
-      } catch (std::bad_cast) {
+      } catch (std::bad_cast&) {
         throw Exceptions::BadCast("Cast from Xpetra::Matrix to Xpetra::CrsMatrixWrap failed");
       }
 #endif

--- a/packages/muelu/src/Utils/MueLu_Utilities_def.hpp
+++ b/packages/muelu/src/Utils/MueLu_Utilities_def.hpp
@@ -177,10 +177,10 @@ namespace MueLu {
       try {
         const Xpetra::EpetraCrsMatrixT<GlobalOrdinal,Node>& tmp_ECrsMtx = dynamic_cast<const Xpetra::EpetraCrsMatrixT<GlobalOrdinal,Node>&>(*crsOp.getCrsMatrix());
         return *tmp_ECrsMtx.getEpetra_CrsMatrix();
-      } catch (std::bad_cast) {
+      } catch (std::bad_cast&) {
         throw Exceptions::BadCast("Cast from Xpetra::CrsMatrix to Xpetra::EpetraCrsMatrix failed");
       }
-    } catch (std::bad_cast) {
+    } catch (std::bad_cast&) {
       throw Exceptions::BadCast("Cast from Xpetra::Matrix to Xpetra::CrsMatrixWrap failed");
     }
   }
@@ -192,10 +192,10 @@ namespace MueLu {
       try {
         Xpetra::EpetraCrsMatrixT<GlobalOrdinal,Node>& tmp_ECrsMtx = dynamic_cast<Xpetra::EpetraCrsMatrixT<GlobalOrdinal,Node>&>(*crsOp.getCrsMatrix());
         return *tmp_ECrsMtx.getEpetra_CrsMatrixNonConst();
-      } catch (std::bad_cast) {
+      } catch (std::bad_cast&) {
         throw Exceptions::BadCast("Cast from Xpetra::CrsMatrix to Xpetra::EpetraCrsMatrix failed");
       }
-    } catch (std::bad_cast) {
+    } catch (std::bad_cast&) {
       throw Exceptions::BadCast("Cast from Xpetra::Matrix to Xpetra::CrsMatrixWrap failed");
     }
   }
@@ -276,10 +276,10 @@ namespace MueLu {
       try {
         const Xpetra::TpetraCrsMatrix<Scalar,LocalOrdinal,GlobalOrdinal,Node>& tmp_ECrsMtx = dynamic_cast<const Xpetra::TpetraCrsMatrix<Scalar,LocalOrdinal,GlobalOrdinal,Node>&>(*crsOp.getCrsMatrix());
         return *tmp_ECrsMtx.getTpetra_CrsMatrix();
-      } catch (std::bad_cast) {
+      } catch (std::bad_cast&) {
         throw Exceptions::BadCast("Cast from Xpetra::CrsMatrix to Xpetra::TpetraCrsMatrix failed");
       }
-    } catch (std::bad_cast) {
+    } catch (std::bad_cast&) {
       throw Exceptions::BadCast("Cast from Xpetra::Matrix to Xpetra::CrsMatrixWrap failed");
     }
   }
@@ -291,10 +291,10 @@ namespace MueLu {
       try {
         Xpetra::TpetraCrsMatrix<Scalar,LocalOrdinal,GlobalOrdinal,Node>& tmp_ECrsMtx = dynamic_cast<Xpetra::TpetraCrsMatrix<Scalar,LocalOrdinal,GlobalOrdinal,Node>&>(*crsOp.getCrsMatrix());
         return *tmp_ECrsMtx.getTpetra_CrsMatrixNonConst();
-      } catch (std::bad_cast) {
+      } catch (std::bad_cast&) {
         throw Exceptions::BadCast("Cast from Xpetra::CrsMatrix to Xpetra::TpetraCrsMatrix failed");
       }
-    } catch (std::bad_cast) {
+    } catch (std::bad_cast&) {
       throw Exceptions::BadCast("Cast from Xpetra::Matrix to Xpetra::CrsMatrixWrap failed");
     }
   }

--- a/packages/muelu/test/scaling/DriverCore.hpp
+++ b/packages/muelu/test/scaling/DriverCore.hpp
@@ -404,7 +404,7 @@ void SystemSolve(Teuchos::RCP<Xpetra::Matrix<Scalar,LocalOrdinal,GlobalOrdinal,N
         ret = solver->solve();
         numIts = solver->getNumIters();
 
-      } catch (std::invalid_argument)
+      } catch (std::invalid_argument&)
 #endif
       {
 

--- a/packages/muelu/test/unit_tests/ParameterList/FactoryFactory.cpp
+++ b/packages/muelu/test/unit_tests/ParameterList/FactoryFactory.cpp
@@ -102,14 +102,14 @@ namespace MueLuTests {
           Teuchos::getValue<std::string>(paramValue);
           RUN;
           continue;
-        } catch (Teuchos::bad_any_cast) { }
+        } catch (Teuchos::bad_any_cast&) { }
 
         const Teuchos::ParameterList& sublist = Teuchos::getValue<Teuchos::ParameterList>(paramValue);
 
         // Test when sublist does not contain type
         try {
           sublist.get<std::string>("type");
-        } catch (Teuchos::Exceptions::InvalidParameterName) {
+        } catch (Teuchos::Exceptions::InvalidParameterName&) {
           RUN;
           continue;
         }

